src/Uno.UI/UI/Xaml/UIElement.RoutedEvents.cs

```diff
@@ -763,23 +696,65 @@ internal bool RaiseEvent(RoutedEvent routedEvent, RoutedEventArgs args, Bubbling
 			return RaiseOnParent(routedEvent, args, parent, ctx);
 		}
 
-		private static void TrackKeyState(RoutedEvent routedEvent, RoutedEventArgs args)
+		/// <summary>
+		/// Raise a tunneling routed event starting from the root and tunneling down to this element.
+		/// </summary>
+		internal void RaiseTunnelingEvent(RoutedEvent routedEvent, RoutedEventArgs args)
 		{
-			if (args is KeyRoutedEventArgs keyArgs)
+			if (!routedEvent.Flag.IsTunnelingEvent())
 			{
-				if (routedEvent == KeyDownEvent)
+				throw new InvalidOperationException($"Event {routedEvent.Name} is not marked as a tunneling event.");
+			}
+
+			// TODO: This is just temporary workaround before proper
+			// keyboard event infrastructure is implemented everywhere
+			// (issue #6074)
+			// The key states will be tracked again in an accompanying bubbling event (e.g. KeyDown for PreviewKeyDown),
+			// but this is fine, since it's idempotent.
+			if (routedEvent.IsKeyEvent)
+			{
+				TrackKeyState(routedEvent, args);
+			}
+
+			// On WinUI, if one of the event handlers reparents this element, the tunneling still goes through the
+			// original path.
+			foreach (var p in this.GetAllParents().Reverse())
+			{
+				if (p is not UIElement parent)
 				{
-					KeyboardStateTracker.OnKeyDown(keyArgs.OriginalKey);
+					break;
 				}
-				else if (routedEvent == KeyUpEvent)
+
+				if (parent._eventHandlerStore.TryGetValue(routedEvent, out var handlers))
 				{
-					KeyboardStateTracker.OnKeyUp(keyArgs.OriginalKey);
+					foreach (var handler in handlers.ToArray())
```

Comment:
As I think that most of the time we have only one handler, we should avoid array creation, but lets keep it for later https://github.com/unoplatform/uno/issues/16320